### PR TITLE
Shout out cooldowns and moderation logs error logs

### LINKED ISSUE #38: Shout out/Go-live cooldowns
# Feature
Users will receive cooldowns for shout outs and go-live posts when streams go offline.

## Acceptance Criteria
- [x] Users shall be put on a shout out/go-live cooldown when their streams go offline.
- [x] Users who are on a shout out/go-live cooldown shall not trigger additional shout outs or go-live posts.
- [x] A user's cooldown will last for 10 minutes starting when their stream last went offline.

## Additional Context
Sushi Bot currently posts these when a stream crashes and then goes live again, which potentially inconveniences server members by  pinging them excessively.

### LINKED ISSUE #39: Moderation logs error handler
# Feature
An extension to the bot’s error handlers, specifically for moderation logs.

## Acceptance Criteria
- [x] The error handler shall handle all errors related to moderation logs.
- [x] The error handler shall output error logs to a designated error log channel in Sushi Hub.
- [x] The error log shall include the following:
  - [x] the moderation log category,
  - [x] the server the error occurred in, and
  - [x] the user who triggered it.

### PR DESCRIPTION
**Pull request by MirageAegis**
# What Does This PR Do?
This PR adds a cooldown to automatic shout outs- and go-live posts, and adds error logs for moderation logs.

## Acceptance Criteria Met
- Users are put on a 10 minute shout out/go-live cooldown when their stream goes offline.
- Users who are on a shout out/go-live cooldown will not trigger any shout outs/go-live posts.
- Moderation logs errors are handled by it own error handler.
- The moderation logs error handler outputs error logs to a designated error log channel in Sushi Hub.
- The error logs include:
  - the moderation log type,
  - the server the error occurred in,
  - the user who triggered it, and
  - a payload if applicable.

## Related Issues
Closes #38 
Closes #39

### CODE REVIEW

File: src/events/shoutout.ts [hunk 1]
```diff
@@ -111,3 +114,56 @@ export const formatGoLivePost = (activity: Activity, template: string): string |
...
+ * Checks whether a user is on cooldown or not.
+ * 
+ * @param user the ID of the user to check
+ */
+export const onCooldown = (user: Snowflake): boolean => {
+    // Get the existing cooldown if there is one
+    const cooldown: NodeJS.Timeout = cooldownCache.get(user);
+
+    return cooldown ? true : false;
+};
```
Comment (MirageAegis) [inline src/events/shoutout.ts:167]: This is the shout out/go-live cooldown logic, which is similar to the cache for server documents.

File: src/events/events.ts [hunk 2]
```diff
@@ -394,8 +395,14 @@ export const onPresenceUpdate = async (client: Client, before: Presence, after:
...
+    // If a stream has ended, add a cooldown and skip
+    if (oldStreams.length > streams.length) {
+        startCooldown(uid);
+        return;
+    }
+
+    // Skip if the member is on cooldown or no stream was found
+    if (onCooldown(uid) || !streams.length) {
         return;
     }
```
Comment (MirageAegis) [inline src/events/events.ts:407]: This is where cooldowns are applied and checked. The checks happen after the application, so cooldowns can be renewed.

File: src/util/error-handler.ts [hunk 3]
```diff
@@ -151,3 +151,146 @@ export const reactionRolesErrorHandler: ErrorHandler = async (ctx: ButtonInterac
...
+ * @param err the error that occurred
+ * @param payload the error payload
+ */
+export const moderationLogsErrorHandler = async <T extends LogsAction>(
+    action: T,
+    server: Guild,
+    trigger: Trigger<T>,
+    err: Error,
+    ...payload: Payload<T>
+): Promise<void> => {
```
Comment (MirageAegis) [inline src/util/error-handler.ts]: This is the error handler for moderation logs errors. It uses conditional types for its parameters.

File: src/util/error-handler.ts [hunk 4]
```diff
@@ -151,3 +151,132 @@ export const reactionRolesErrorHandler: ErrorHandler = async (ctx: ButtonInterac
...
+ * @param trigger the member or user who triggered the error
+ * @param err the error that occurred
+ * @param payload the error payload
+ */
+export const moderationLogsErrorHandler = async <T extends LogsAction>(
+    action: T,
+    server: Guild,
+    trigger: User,
+    err: Error,
+    ...payload: Payload<T>
```
Comment (MirageAegis) [inline src/util/error-handler.ts:216]: This is the error handler for moderation log errors. It uses a conditional type for its payload, which depends on the moderations log action.

File: src/util/error-handler.ts [hunk 5]
```diff
@@ -151,3 +151,132 @@ export const reactionRolesErrorHandler: ErrorHandler = async (ctx: ButtonInterac
...
+    UserUpdate = "User Update"
+}
+
+/**
+ * Actions related to the messages logs category.
+ */
+export enum MessageLogsAction {
+    MessageEdit = "Message Edit",
+    MessageDelete = "Message Delete"
+}
```
Comment (MirageAegis) [inline src/util/error-handler.ts:179]: These are the enums for each moderation log action. They are split into the 3 moderation logs categories.

File: src/util/error-handler.ts [hunk 6]
```diff
@@ -151,3 +151,132 @@ export const reactionRolesErrorHandler: ErrorHandler = async (ctx: ButtonInterac
...
+ * The before and after states of a Message if it's a message edit action, or
+ * a single message if it's a message delete action, or
+ * the before and after states of a User if it's a user update action, or
+ * the before and after states of a Member if it's a member update action.
+ */
+type Payload<T extends LogsAction> = T extends MessageLogsAction.MessageEdit ? [before: Message, after: Message] :
+    T extends MessageLogsAction ? [message: Message] :
+    T extends ProfileLogsAction.UserUpdate ? [before: User, after: User] :
+    T extends ProfileLogsAction.MemberUpdate ? [before: GuildMember, after: GuildMember] :
+    [];
```
Comment (MirageAegis) [inline src/util/error-handler.ts:199]: The `payload`'s type depends on what data can be collected from the error. For message edits, 2 message objects will be collected; for user profile updates, 2 user objects will be collected; etc.

File: src/util/error-handler.ts [hunk 7]
```diff
@@ -151,3 +151,132 @@ export const reactionRolesErrorHandler: ErrorHandler = async (ctx: ButtonInterac
...
+                           `Server: ${server?.name}\n` +
+                           `Trigger: ${trigger.id} (${trigger.username})\n\n` +
+                           `${before ? "Payload\n\n" : ""}` +
+                           `${before ? before + "\n": "" }${after ? after + "\n" : ""}` +
+                           `${before ? "\n" : ""}` +
+                           // Error name, and error code if it's a Discord API error
+                           `${err.name}${err instanceof DiscordAPIError ? `: ${err.code} (${err.message})` : ""}\n` +
+                           "```";
+    
+    await getAdminLogsChannel().send(report);
```
Comment (MirageAegis) [inline src/util/error-handler.ts:281]: The error data is then logged to the error log.

**Review by MirageAegis — COMMENTED** [review]
This review will walk through all the relevant changes this PR brings.

**Review by zahatikoff — APPROVED** [review]

**Pull request merged**